modules/storage/src/StockStorageAPI.php

```diff
@@ -21,180 +21,200 @@ public function createTransaction($variation_id, $location_id, $zone, $quantity,
       ->fields('ill')
-      ->condition('location_id', $location_id, '=')
+      ->condition('location_id', $location_id)
       ->condition('variation_id', $variation_id)
       ->execute()
       ->fetch();
-    if ($result) {
-      $stock_level = $result->qty;
-      $last_transaction = $result->last_transaction_id;
+
+    return [
+      'qty' => $result ? $result->qty : 0,
+      'last_transaction' => $result ? $result->last_transaction_id : 0
+    ];
+  }
+
+  /**
+   * Sets the stock level and last transaction for a given location and variation.
+   *
+   * Creates first stock level transaction record if none exists.
+   *
+   * @param int $location_id
+   *   Location id.
+   *
+   * @param int $variation_id
+   *   Variation id.
+   *
+   * @param int $qty
+   *   Quantity.
+   *
+   * @param int $last_txn
+   *   Last transaction id.
+   */
+  public function setLocationStockLevel($location_id, $variation_id, $qty, $last_txn) {
+    $db = \Drupal::database();
+    $existing = $db->select('cs_inventory_location_level', 'ill')
+      ->fields('ill')
+      ->condition('location_id', $location_id)
+      ->condition('variation_id', $variation_id)
+      ->execute();
+    if ($existing) {
+      $db->update('cs_inventory_location_level')
+        ->fields([
+          'qty' => $qty,
+          'last_transaction_id' => $last_txn,
+        ])
+        ->condition('location_id', $location_id, '=')
+        ->condition('variation_id', $variation_id, '=')
+        ->execute();
     }
     else {
-      $stock_level = 0;
-      $last_transaction = 0;
-      // Create a record.
-      // @todo - Deprecated replace with https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Database%21Connection.php/function/Connection%3A%3Ainsert/8
-      $io = db_insert('cs_inventory_location_level');
-      $io->fields([
-        'location_id',
-        'variation_id',
-        'qty',
-        'last_transaction_id',
-      ]);
-      $io->values([$location_id, $variation_id, 0, 0]);
-      $io->execute();
+      $db->insert('cs_inventory_location_level')
+        ->fields(['location_id', 'variation_id', 'qty', 'last_transaction_id'])
+        ->values([$location_id, $variation_id, 0, 0])
+        ->execute();
     }
+  }
 
-    // Get the last transaction id.
-    // @todo - need to use a higher level method.
-    $query = "select max(trid) as max_id from `{cs_inventory_transaction}`
-      WHERE (`location_id` = '" . $location_id . "')
-      AND (`variation_id` =" . $variation_id . ")
-      GROUP BY location_id";
-    $result = $db->query($query)->fetch();
-    if (!$result) {
-      // If no new transaction then nothing to do.
-      return;
-    }
-    $max_transaction = $result->max_id;
-
-    // Total all transactions between last and max transactions.
-    // @todo - need to use a higher level method.
-    $query = "SELECT location_id, sum(qty) as transactions_qty FROM `{cs_inventory_transaction}`
```

Comment:
That's great we only need to make sure it is also used in getLocationStockTransactionSum(see later comment)